pkg/controller/istiocsr/client.go

```diff
@@ -0,0 +1,115 @@
+package istiocsr
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+type ctrlClientImpl struct {
+	client.Client
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate -o fakes . ctrlClient
+type ctrlClient interface {
+	Get(context.Context, client.ObjectKey, client.Object) error
+	List(context.Context, client.ObjectList, ...client.ListOption) error
+	StatusUpdate(context.Context, client.Object, ...client.SubResourceUpdateOption) error
+	Update(context.Context, client.Object, ...client.UpdateOption) error
+	UpdateWithRetry(context.Context, client.Object, ...client.UpdateOption) error
```

Comment:
Any particular reason why `UpdateWithRetry` is used in all places instead of only `Update`? If the update fails the controller should ideally reconcile the object again. Normally, the `RetryOnConflict` function is used in places where there's no retry code is inherently available.

Reply:
`UpdateWithRetry` makes use of `RetryOnConflict` whenever resource update fails with `ResourceVersion` conflict error. The approach is to avoid whole reconciliation happening on all resources when write fails with conflict error for few of the resources. Worst case  scenario, write fails with conflict error for all resources, but still retry is better than being in multiple loops when it write keeps failing with conflict error.

Reply:
Are these objects also being updated from some other place? If not, the conflict error will not come up and in which case the `Update` function should suffice.